### ISSUE TITLE
Hillshade renderer artifacts

### ISSUE THREAD

**Issue by qgib** [qgis/QGIS#23859]
Author Name: **Paolo Cavallini** (@pcav)
Original Redmine Issue: [15943](https://issues.qgis.org/issues/15943)
Affected QGIS version: 3.0.0
Redmine category:symbology

---
Borders of cells are visible. See attached.
Possibly more apparent in master, but present since the beginning.

---

- [hillshade.png](https://issues.qgis.org/attachments/download/10576/hillshade.png) (Paolo Cavallini)
- [nn vs bilinear.png](https://issues.qgis.org/attachments/download/11926/nn%20vs%20bilinear.png) (Johannes Kroeger)

**Comment by AlisterH** [comment]
Note that there is a similar effect _around the edge of the canvas_ if your raster is in a different projection to your canvas, and is being reprojected "on the fly".  This one is _more_ noticeable if smoothing is enabled.  Presumably it is just because the OTF reprojection code does not use values from outside the map canvas in its calculations.

For example:
![image](https://user-images.githubusercontent.com/714600/61196790-1fc2cf00-a725-11e9-9aab-dc6ce122a583.png)

If you get the zoom level right it looks pretty extreme:
![image](https://user-images.githubusercontent.com/714600/61196624-0e2cf780-a724-11e9-9b01-9124e0dcff1b.png)

**Comment by AlisterH** [comment]
Sorry, I was confused in my last comment - this issue is apparent when the raster is in the same projection as the canvas - it is _not_ apparent if it is being reprojected "on the fly".  If smoothing is turned off it is only noticeably if you zoom out.

**Comment by nirvn** [comment]
This is one of the many limitation of the hillshade renderer. If I were you, I'd create an hillshade raster from your DEM using processing, I doubt the renderer will be improved on the short term.

**Comment by bellanz80** [comment]
Hi, i have a DEM. 
DEM visualized with Hillshade style. Resampling: "Bilineal"
In QGIS 3.8 and 3.10, when changing the Resamplig to "bilinear", it stays the same as "nearest", where the pixels are visible. Meanwhile, in QGIS 2.18 and 3.4 it works correctly.

![Presentación1](https://user-images.githubusercontent.com/38449926/73043319-424b4880-3e2a-11ea-8f1c-f75c87d10cbc.png)

[clip_dem.zip](https://github.com/qgis/QGIS/files/4106753/clip_dem.zip)

**Comment by bellanz80** [comment]
Important update. He tested with QGIS 3.10.2 and a clean profile, no problem ... everything worked perfect ... will the plugins be? impossible now to know, I have many installed ..

**Comment by AlisterH** [comment]
I guess it is easy enough to try copying all the plugins into the clean profile, enabling them all and seeing if it is OK.
Presumably you did confirm it was a problem in fresh empty projects, not just in your existing projects?

**Comment by bellanz80** [comment]
thanks. But in reality it is not easy to copy the plug from one profile to another. Only a majority is included in the profile folder.
And yes, they were new projects. It became clear that something subsequently installed stopped working the resampling.

**Comment by bellanz80** [comment]
Update: I managed to copy all the plug of my profile, normal, to a clean profile (I had not seen the python folder inside the profile).
And as expected, the problem continues to appear ... confirmed are the plug ... a day that has a few hours off, I will do a test.
I also reviewed the OPENCL, and it does NOT change the result, activating it or not.
Thank you!!

**Comment by bellanz80** [comment]
Last update: I confirm that the problem is NOT the version of QGIS, but the plugs.
I opened the profile with plug in QGIS 3.4, and the error was generated. Thank you

**Comment by gioman** [comment]
Should we close this?

**Comment by roya0045** [comment]
I guess it would be nice to isolate the plugin that causes this in order to have a future reference or allow the author to fix it.

**Comment by gioman** [comment]
> I guess it would be nice to isolate the plugin that causes this in order to have a future reference or allow the author to fix it.

yes, but if is a plugin to blame this ticket does not belong here anyway.

**Comment by bellanz80** [comment]
hi. After 3 hours of testing, I found the problem and the solution ...
I don't know if it's a bug, or if it's completely normal, but the "render resamplig" stopped working because the CRS of the layer is UTM (EPSG 32615), but the CRS of the project was Geographic (EPSG 4326).
By putting the two systems the same, the resample works without problem.
With your great experience, is it a bug, or is it all normal and I was wrong the process? Thank you

**Comment by nirvn** [comment]
It's a known limitation. We can safely close this as the limitation is well known.

**Comment by pcav** [comment]
Sorry, I don't agree: even if known, this is a real issue for the user. Apparently it should be easily overcome by activating resampling.

**Comment by jfbourdon** [comment]
If I may add my voice, I'm glad to have came across this issue as it was a real bummer for me to see that the hillshade renderer was not usable due to the cells being visible. Only now, after 3 years of using QGIS, I learn that:

1. The CRS of the layer needs to be the same of the one of the project to minimize artifacts;
2. The sampling method must be changed from the default *nearest* to *bilinear* (and *average* for zoom out).

It would be nice if these info were written on the Hillshade panel or a warning issued when selecting the hillshade renderer.

**Comment by AlisterH** [comment]
> hi. After 3 hours of testing, I found the problem and the solution ...
> I don't know if it's a bug, or if it's completely normal, but the "render resamplig" stopped working because the CRS of the layer is UTM (EPSG 32615), but the CRS of the project was Geographic (EPSG 4326).
> By putting the two systems the same, the resample works without problem.
> With your great experience, is it a bug, or is it all normal and I was wrong the process? Thank you

Wow, I never realised this, but haven't you have described it backwards?  Testing here it seems "The resample works without problem" when the layer is reprojected i.e. when the layer CRS and project CRS are DIFFERENT.

**Comment by AlisterH** [comment]
> The sampling method must be changed from the default nearest to bilinear (and average for zoom out).

I avoid this annoyance by setting these as default in Settings>Options>Rendering>Rasters.  But it probably isn't a great idea for layers that aren't hillshaded.

**Comment by AlisterH** [comment]
Sorry, I don't agree: even if known, this is a real issue for the user. Apparently it should be easily overcome by activating resampling.

> Sorry, I don't agree: even if known, this is a real issue for the user. Apparently it should be easily overcome by activating resampling.

What do you mean?  Changing the default setting as per my last comment?  Always resampling hillshaded layers regardless of the layer's setting?  Automatically changing the layer's setting when hillshading is enabled?

**Comment by jfbourdon** [comment]
> I avoid this annoyance by setting these as default in Settings>Options>Rendering>Rasters. But it probably isn't a great idea for layers that aren't hillshaded.

As you correctly pointed out, changing the global default value isn't a great solution because if you have a classified raster, you wouldn't want a bilinear resampling, so it's just replacing a annoyance by another.

Could the current state of resampling be saved when choosing the hillshade renderer so that when the user switch back to *Singleband pseudocolor* for example, it automatically revert resampling to whatever it was before? That way the hillshade renderer could be set to always be bilinear without causing issues.

>Haven't you have described it backwards? Testing here it seems "The resample works without problem" when the layer is reprojected i.e. when the layer CRS and project CRS are DIFFERENT.

What I observe is effectively the opposite: if the layer CRS and project CRS are DIFFERENT, the issue is present, even with bilinear resampling, but less present than with nearest neighbour.

**Comment by Lorchi74** [comment]
I confirm this rendering bug with QGIS version 3.34.8. Both the layer and the project have the same CRS: EPSG:25832 (UTM 32N).
